src/allmydata/storage/http_common.py

```diff
@@ -34,3 +38,14 @@ class Secrets(Enum):
     LEASE_RENEW = "lease-renew-secret"
     LEASE_CANCEL = "lease-cancel-secret"
     UPLOAD = "upload-secret"
+
+
+def get_spki_hash(certificate: Certificate) -> bytes:
+    """
+    Get the public key hash, as per RFC 7469: base64 of sha256 of the public
+    key encoded in DER + Subject Public Key Info format.
+    """
+    public_key_bytes = certificate.public_key().public_bytes(
+        Encoding.DER, PublicFormat.SubjectPublicKeyInfo
+    )
+    return b64encode(sha256(public_key_bytes).digest()).strip().rstrip(b"=")
```

Comment:
The GBS spec calls for "urlsafe-base64 (without padding)" here.  I think I believed that was necessary to be able to represent NURLs using standard URL-shaped tools.  This hash ends up in the userinfo part of the URI and it looks like percent encoding is allowed there (https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1) so it seems like it is _not_ *necessary* to use urlsafe-base64 to play nicely with other tools.

I wonder if it still might be preferable to avoid ever having a hash with percent encoding in.  Maybe it would be slightly annoying to sometimes have to deal with such a hash "by hand" (ie, _sometimes_ naive copy/paste will give you a bogus string and maybe you won't always immediately notice)?  The flip side being that sometimes if you forget to handle the hash using the right base64 flavor you will have decoding problems.

So ... maybe it is a wash.

Reply:
I guess I'm not using URL-safe base64 variant here, at minimum... should I do that, or do you want to switch to % encoding instead?